Unsupported borrow kind `Fake(Shallow)`

The following code:
```Rust
fn do_stuff(x: Option<usize>, check: usize) -> Option<usize> {
    match x {
        Some(i) if i == check => Some(2),
        _ => None,
    }
}
```
Fails with:
```
error[E0999]: unsupported statement
  --> src/main.rs:80:11
   |
80 |     match x {
   |           ^
   |
   = note: unsupported borrow kind `Fake(Shallow)`

```

However the code 
```Rust
fn do_stuff_different(x: Option<usize>, check: usize) -> Option<usize> {
    match x {
        Some(i) => Some(2),
        _ => None,
    }
}
```
does not fail.
The Fake(Shallow) borrow kind appears to only occur when using match arms with extra constraints on the arm.

--- a/crates/flux-middle/src/rty/mod.rs
+++ b/crates/flux-middle/src/rty/mod.rs
@@ -859,7 +859,7 @@ impl Ty {
     }
 
     /// Replace all regions with a [`ReVar`] assigning each a unique [`RegionVid`]. This is used
-    /// to have a unique var identifying each position such that we can infer a region substitution
+    /// to have a unique identifier for each position such that we can infer a region substitution
     /// when assigning a type to a place. This way we can recover the regions in the original rust
     /// type. See `flux_refineck::type_env::TypeEnv::assign`
     pub fn replace_regions_with_unique_vars(&self) -> Ty {

--- a/crates/flux-middle/src/rty/subst.rs
+++ b/crates/flux-middle/src/rty/subst.rs
@@ -12,11 +12,11 @@ use super::{
 };
 use crate::rty::*;
 
-pub fn match_regions(ty1: &Ty, ty2: &rustc::ty::Ty) -> Ty {
-    let ty1 = ty1.replace_regions_with_unique_vars();
+pub fn match_regions(a: &Ty, b: &rustc::ty::Ty) -> Ty {
+    let a = a.replace_regions_with_unique_vars();
     let mut subst = RegionSubst::default();
-    subst.infer_from_ty(&ty1, ty2);
-    subst.apply(&ty1)
+    subst.infer_from_ty(&a, b);
+    subst.apply(&a)
 }
 
 #[derive(Default, Debug)]
@@ -41,52 +41,41 @@ impl RegionSubst {
         t.fold_with(&mut Folder(self))
     }
 
-    fn infer_from_ty(&mut self, ty1: &Ty, ty2: &rustc::ty::Ty) {
+    fn infer_from_ty(&mut self, a: &Ty, b: &rustc::ty::Ty) {
         use rustc::ty;
-        match (ty1.kind(), ty2.kind()) {
-            (TyKind::Exists(ty1), _) => {
-                self.infer_from_ty(ty1.as_ref().skip_binder(), ty2);
+        match (a.kind(), b.kind()) {
+            (TyKind::Exists(ty_a), _) => {
+                self.infer_from_ty(ty_a.as_ref().skip_binder(), b);
             }
-            (TyKind::Constr(_, ty1), _) => {
-                self.infer_from_ty(ty1, ty2);
+            (TyKind::Constr(_, ty_a), _) => {
+                self.infer_from_ty(ty_a, b);
             }
-            (TyKind::Indexed(bty, _), _) => {
-                self.infer_from_bty(bty, ty2);
+            (TyKind::Indexed(bty_a, _), _) => {
+                self.infer_from_bty(bty_a, b);
             }
-            (TyKind::Ptr(PtrKind::Mut(r1), _), ty::TyKind::Ref(r2, _, mutbl)) => {
+            (TyKind::Ptr(PtrKind::Mut(re_a), _), ty::TyKind::Ref(re_b, _, mutbl)) => {
                 debug_assert!(mutbl.is_mut());
-                self.infer_from_region(*r1, *r2);
+                self.infer_from_region(*re_a, *re_b);
             }
-            (TyKind::StrgRef(r1, ..), ty::TyKind::Ref(r2, _, mutbl)) => {
+            (TyKind::StrgRef(re_a, ..), ty::TyKind::Ref(re_b, _, mutbl)) => {
                 debug_assert!(mutbl.is_mut());
-                self.infer_from_region(*r1, *r2);
+                self.infer_from_region(*re_a, *re_b);
             }
             _ => {}
         }
     }
 
-    fn infer_from_bty(&mut self, bty: &BaseTy, ty: &rustc::ty::Ty) {
+    fn infer_from_bty(&mut self, a: &BaseTy, ty: &rustc::ty::Ty) {
         use rustc::ty;
-        match (bty, ty.kind()) {
+        match (a, ty.kind()) {
             (BaseTy::Ref(r1, ty1, _), ty::TyKind::Ref(r2, ty2, _)) => {
                 self.infer_from_region(*r1, *r2);
                 self.infer_from_ty(ty1, ty2);
             }
-            (BaseTy::Adt(_, args1), ty::TyKind::Adt(_, args2)) => {
-                debug_assert_eq!(args1.len(), args2.len());
-                for (arg1, arg2) in iter::zip(args1, args2) {
-                    match (arg1, arg2) {
-                        (GenericArg::Base(ctor1), ty::GenericArg::Ty(ty2)) => {
-                            self.infer_from_bty(ctor1.as_bty_skipping_binder(), ty2);
-                        }
-                        (GenericArg::Ty(ty1), ty::GenericArg::Ty(ty2)) => {
-                            self.infer_from_ty(ty1, ty2);
-                        }
-                        (GenericArg::Lifetime(re1), ty::GenericArg::Lifetime(re2)) => {
-                            self.infer_from_region(*re1, *re2);
-                        }
-                        _ => {}
-                    }
+            (BaseTy::Adt(_, args_a), ty::TyKind::Adt(_, args_b)) => {
+                debug_assert_eq!(args_a.len(), args_b.len());
+                for (arg_a, arg_b) in iter::zip(args_a, args_b) {
+                    self.infer_from_generic_arg(arg_a, arg_b);
                 }
             }
             (BaseTy::Tuple(fields1), ty::TyKind::Tuple(fields2)) => {
@@ -99,21 +88,32 @@ impl RegionSubst {
         }
     }
 
-    fn infer_from_region(&mut self, re1: Region, re2: Region) {
-        let ReVar(var) = re1 else { return };
+    fn infer_from_generic_arg(&mut self, a: &GenericArg, b: &rustc::ty::GenericArg) {
+        use rustc::ty;
+        match (a, b) {
+            (GenericArg::Base(ctor_a), ty::GenericArg::Ty(ty_b)) => {
+                self.infer_from_bty(ctor_a.as_bty_skipping_binder(), ty_b);
+            }
+            (GenericArg::Ty(ty_a), ty::GenericArg::Ty(ty_b)) => {
+                self.infer_from_ty(ty_a, ty_b);
+            }
+            (GenericArg::Lifetime(re_a), ty::GenericArg::Lifetime(re_b)) => {
+                self.infer_from_region(*re_a, *re_b);
+            }
+            _ => {}
+        }
+    }
+
+    fn infer_from_region(&mut self, a: Region, b: Region) {
+        let ReVar(var) = a else { return };
         match self.map.entry(var) {
             hash_map::Entry::Occupied(entry) => {
-                if entry.get() != &re2 {
-                    bug!(
-                        "ambiguous region substitution: {:?} -> [{:?}, {:?}]",
-                        re1,
-                        entry.get(),
-                        re2
-                    );
+                if entry.get() != &b {
+                    bug!("ambiguous region substitution: {:?} -> [{:?}, {:?}]", a, entry.get(), b);
                 }
             }
             hash_map::Entry::Vacant(entry) => {
-                entry.insert(re2);
+                entry.insert(b);
             }
         }
     }

--- a/crates/flux-middle/src/rustc/lowering.rs
+++ b/crates/flux-middle/src/rustc/lowering.rs
@@ -19,9 +19,9 @@ use rustc_trait_selection::traits::SelectionContext;
 
 use super::{
     mir::{
-        replicate_infer_ctxt, AggregateKind, AssertKind, BasicBlockData, BinOp, Body, BorrowKind,
-        CallArgs, CastKind, Constant, FakeReadCause, LocalDecl, Operand, Place, PlaceElem,
-        PointerCast, Rvalue, Statement, StatementKind, Terminator, TerminatorKind,
+        replicate_infer_ctxt, AggregateKind, AssertKind, BasicBlockData, BinOp, Body, CallArgs,
+        CastKind, Constant, LocalDecl, Operand, Place, PlaceElem, PointerCast, Rvalue, Statement,
+        StatementKind, Terminator, TerminatorKind,
     },
     ty::{
         AdtDef, AdtDefData, AliasKind, Binder, BoundRegion, BoundVariableKind, Clause, ClauseKind,
@@ -216,9 +216,7 @@ impl<'sess, 'tcx> LoweringCtxt<'_, 'sess, 'tcx> {
             }
             rustc_mir::StatementKind::FakeRead(box (cause, place)) => {
                 StatementKind::FakeRead(Box::new((
-                    self.lower_fake_read_cause(*cause)
-                        .ok_or_else(|| errors::UnsupportedMir::from(stmt))
-                        .emit(self.sess)?,
+                    *cause,
                     lower_place(place)
                         .map_err(|reason| errors::UnsupportedMir::statement(span, reason))
                         .emit(self.sess)?,
@@ -255,18 +253,6 @@ impl<'sess, 'tcx> LoweringCtxt<'_, 'sess, 'tcx> {
             }
         };
         Ok(Statement { kind, source_info: stmt.source_info })
-    }
-
-    fn lower_fake_read_cause(&self, cause: rustc_mir::FakeReadCause) -> Option<FakeReadCause> {
-        match cause {
-            rustc_mir::FakeReadCause::ForLet(def_id) => Some(FakeReadCause::ForLet(def_id)),
-            rustc_mir::FakeReadCause::ForMatchedPlace(def_id) => {
-                Some(FakeReadCause::ForMatchedPlace(def_id))
-            }
-            rustc_mir::FakeReadCause::ForMatchGuard
-            | rustc_mir::FakeReadCause::ForGuardBinding
-            | rustc_mir::FakeReadCause::ForIndex { .. } => None,
-        }
     }
 
     fn lower_terminator(
@@ -407,11 +393,7 @@ impl<'sess, 'tcx> LoweringCtxt<'_, 'sess, 'tcx> {
                 ))
             }
             rustc_mir::Rvalue::Ref(region, bk, p) => {
-                Ok(Rvalue::Ref(
-                    lower_region(region)?,
-                    self.lower_borrow_kind(*bk)?,
-                    lower_place(p)?,
-                ))
+                Ok(Rvalue::Ref(lower_region(region)?, *bk, lower_place(p)?))
             }
             rustc_mir::Rvalue::UnaryOp(un_op, op) => {
                 Ok(Rvalue::UnaryOp(*un_op, self.lower_operand(op)?))
@@ -442,19 +424,6 @@ impl<'sess, 'tcx> LoweringCtxt<'_, 'sess, 'tcx> {
             | rustc_mir::Rvalue::CopyForDeref(_)
             | rustc_mir::Rvalue::ShallowInitBox(_, _) => {
                 Err(UnsupportedReason::new(format!("unsupported rvalue `{rvalue:?}`")))
-            }
-        }
-    }
-
-    fn lower_borrow_kind(
-        &self,
-        bk: rustc_mir::BorrowKind,
-    ) -> Result<BorrowKind, UnsupportedReason> {
-        match bk {
-            rustc_mir::BorrowKind::Shared => Ok(BorrowKind::Shared),
-            rustc_mir::BorrowKind::Mut { kind } => Ok(BorrowKind::Mut { kind }),
-            rustc_mir::BorrowKind::Fake(_) => {
-                Err(UnsupportedReason::new(format!("unsupported borrow kind `{bk:?}`")))
             }
         }
     }

--- a/crates/flux-middle/src/rustc/mir.rs
+++ b/crates/flux-middle/src/rustc/mir.rs
@@ -10,18 +10,18 @@ use itertools::Itertools;
 pub use rustc_borrowck::borrow_set::BorrowData;
 use rustc_borrowck::consumers::{BodyWithBorrowckFacts, BorrowIndex};
 use rustc_data_structures::{fx::FxIndexMap, graph::dominators::Dominators};
-use rustc_hir::def_id::{DefId, LocalDefId};
+use rustc_hir::def_id::DefId;
 use rustc_index::IndexSlice;
 use rustc_infer::infer::TyCtxtInferExt;
 use rustc_macros::{Decodable, Encodable};
 use rustc_middle::{
-    mir::{self, MutBorrowKind},
+    mir,
     ty::{FloatTy, IntTy, ParamConst, TyCtxt, UintTy},
 };
 pub use rustc_middle::{
     mir::{
-        BasicBlock, Local, LocalKind, Location, SourceInfo, SwitchTargets, UnOp, UnwindAction,
-        RETURN_PLACE, START_BLOCK,
+        BasicBlock, BorrowKind, FakeBorrowKind, FakeReadCause, Local, LocalKind, Location,
+        SourceInfo, SwitchTargets, UnOp, UnwindAction, RETURN_PLACE, START_BLOCK,
     },
     ty::{UserTypeAnnotationIndex, Variance},
 };
@@ -195,11 +195,6 @@ pub enum Rvalue {
     Repeat(Operand, Const),
 }
 
-pub enum BorrowKind {
-    Mut { kind: MutBorrowKind },
-    Shared,
-}
-
 #[derive(Copy, Clone)]
 pub enum CastKind {
     IntToInt,
@@ -287,11 +282,6 @@ pub enum Constant {
     Param(ParamConst, Ty),
     /// General catch-all for constants of a given Ty
     Opaque(Ty),
-}
-
-pub enum FakeReadCause {
-    ForLet(Option<LocalDefId>),
-    ForMatchedPlace(Option<LocalDefId>),
 }
 
 impl<'tcx> Terminator<'tcx> {
@@ -647,6 +637,12 @@ impl fmt::Debug for Rvalue {
             Rvalue::Ref(r, BorrowKind::Shared, place) => {
                 write!(f, "&{} {place:?}", region_to_string(*r))
             }
+            Rvalue::Ref(r, BorrowKind::Fake(FakeBorrowKind::Shallow), place) => {
+                write!(f, "&{} fake shallow {place:?}", region_to_string(*r))
+            }
+            Rvalue::Ref(r, BorrowKind::Fake(FakeBorrowKind::Deep), place) => {
+                write!(f, "&{} fake deep {place:?}", region_to_string(*r))
+            }
             Rvalue::Discriminant(place) => write!(f, "discriminant({place:?})"),
             Rvalue::BinaryOp(bin_op, op1, op2) => write!(f, "{bin_op:?}({op1:?}, {op2:?})"),
             Rvalue::CheckedBinaryOp(bin_op, op1, op2) => {
@@ -742,15 +738,6 @@ impl fmt::Debug for Constant {
             Constant::Char => write!(f, "\"<opaque char>\""),
             Constant::Opaque(ty) => write!(f, "<opaque {:?}>", ty),
             Constant::Param(p, _) => write!(f, "{:?}", p),
-        }
-    }
-}
-
-impl fmt::Debug for FakeReadCause {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            FakeReadCause::ForLet(def_id) => write!(f, "ForLet({def_id:?})"),
-            FakeReadCause::ForMatchedPlace(def_id) => write!(f, "ForMatchedPlace({def_id:?})"),
         }
     }
 }

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -784,7 +784,7 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
                 env.borrow(self.genv, rcx, *r, Mutability::Mut, place)
                     .with_span(stmt_span)
             }
-            Rvalue::Ref(r, BorrowKind::Shared, place) => {
+            Rvalue::Ref(r, BorrowKind::Shared | BorrowKind::Fake(..), place) => {
                 env.borrow(self.genv, rcx, *r, Mutability::Not, place)
                     .with_span(stmt_span)
             }

--- a/crates/flux-refineck/src/ghost_statements/fold_unfold.rs
+++ b/crates/flux-refineck/src/ghost_statements/fold_unfold.rs
@@ -9,8 +9,9 @@ use flux_middle::{
     rty,
     rustc::{
         mir::{
-            BasicBlock, Body, FieldIdx, Local, Location, Operand, Place, PlaceElem, Rvalue,
-            Statement, StatementKind, Terminator, TerminatorKind, VariantIdx, FIRST_VARIANT,
+            BasicBlock, Body, BorrowKind, FieldIdx, Local, Location, Operand, Place, PlaceElem,
+            Rvalue, Statement, StatementKind, Terminator, TerminatorKind, VariantIdx,
+            FIRST_VARIANT,
         },
         ty::{AdtDef, GenericArgs, Ty, TyKind},
     },
@@ -304,8 +305,11 @@ impl<'a, 'genv, 'tcx, M: Mode> FoldUnfoldAnalysis<'a, 'genv, 'tcx, M> {
                     Rvalue::Use(op) | Rvalue::Cast(_, op, _) | Rvalue::UnaryOp(_, op) => {
                         self.operand(op, env)?;
                     }
-                    Rvalue::Ref(.., place) => {
-                        M::projection(self, env, place, ProjKind::Other)?;
+                    Rvalue::Ref(.., bk, place) => {
+                        // Fake borrows should not cause the place to fold
+                        if !matches!(bk, BorrowKind::Fake(_)) {
+                            M::projection(self, env, place, ProjKind::Other)?;
+                        }
                     }
                     Rvalue::CheckedBinaryOp(_, op1, op2) | Rvalue::BinaryOp(_, op1, op2) => {
                         self.operand(op1, env)?;

--- a/tests/tests/neg/enums/issue-687.rs
+++ b/tests/tests/neg/enums/issue-687.rs
@@ -1,0 +1,7 @@
+#[flux::sig(fn(x: Option<usize{v: check == v => v > 0}>, check: usize) -> Option<usize{v: v > 1}>)]
+fn test00(x: Option<usize>, check: usize) -> Option<usize> {
+    match x {
+        Some(i) if i == check => Some(i), //~ERROR refinement type error
+        _ => None,
+    }
+}

--- a/tests/tests/pos/surface/issue-687.rs
+++ b/tests/tests/pos/surface/issue-687.rs
@@ -1,0 +1,7 @@
+#[flux::sig(fn(x: Option<usize{v: check == v => v > 0}>, check: usize) -> Option<usize{v: v > 0}>)]
+fn test00(x: Option<usize>, check: usize) -> Option<usize> {
+    match x {
+        Some(i) if i == check => Some(i),
+        _ => None,
+    }
+}

